crates/nix_rs/src/store.rs

```diff
@@ -147,38 +147,33 @@ impl NixStoreCmd {
         &self,
         out_paths: HashSet<StorePath>,
     ) -> Result<HashSet<StorePath>, NixStoreCmdError> {
-        let mut all_drvs = HashSet::new();
-        for out in out_paths.iter() {
-            let drv = self.nix_store_query_deriver(out.clone()).await?;
-            all_drvs.insert(drv);
-        }
-        let mut all_outs = HashSet::new();
-        for drv in all_drvs {
-            let deps = self.nix_store_query_requisites_with_outputs(drv).await?;
-            all_outs.extend(deps.into_iter());
-        }
-        Ok(all_outs)
+        let all_drvs = self
+            .nix_store_query_deriver(&out_paths.iter().cloned().collect::<Vec<_>>())
+            .await?;
+        let all_outs = self
+            .nix_store_query_requisites_with_outputs(&all_drvs)
+            .await?;
+        Ok(all_outs.into_iter().collect())
     }
 
-    /// Return the derivation used to build the given build output.
+    /// Return the derivations used to build the given build output.
     pub async fn nix_store_query_deriver(
         &self,
-        out_path: StorePath,
-    ) -> Result<PathBuf, NixStoreCmdError> {
+        out_paths: &[StorePath],
+    ) -> Result<Vec<PathBuf>, NixStoreCmdError> {
         let mut cmd = self.command();
-        cmd.args([
-            "--query",
-            "--valid-derivers",
-            out_path.as_path().to_string_lossy().as_ref(),
-        ]);
+        cmd.args(["--query", "--valid-derivers"])
+            .args(out_paths.iter().map(StorePath::as_path));
+
         crate::command::trace_cmd(&cmd);
         let out = cmd.output().await?;
         if out.status.success() {
-            let drv_path = String::from_utf8(out.stdout)?.trim().to_string();
-            if drv_path == "unknown-deriver" {
+            let out = String::from_utf8(out.stdout)?.trim().to_string();
+            let drv_paths: HashSet<PathBuf> = out.lines().map(PathBuf::from).collect();
```

Comment:
Converting it to HashSet first and later to Vec while returning  as I want this check for `unknown-deriver` in constant time.